src/algorithms/reco/MatchClusters.cc

```diff
@@ -14,197 +15,170 @@
-                m_log->debug("    --> found matching cluster with energy: {}", clus.getEnergy());
-                m_log->debug("    --> adding cluster to reconstructed particle");
-                outpart.addToClusters(clus);
-                clusterMap.erase(mcID);
-            }
-
-            // create truth associations
-            auto assoc = outpartsassoc->create();
-            assoc.setRecID(outpart.getObjectID().index);
-            assoc.setSimID(mcID);
-            assoc.setWeight(1.0);
-            assoc.setRec(outpart);
-            assoc.setSim((*mcparticles)[mcID]);
         }
 
-        // 2. Now loop over all remaining clusters and add neutrals. Also add in Hcal energy
-        // if a matching cluster is available
-        m_log->debug("Step 2/2: Creating neutrals for remaining clusters...");
-        for (const auto &[mcID, clus]: clusterMap) {
-            m_log->debug(" --> Processing unmatched cluster with energy: {}", clus.getEnergy());
-
-
-            // get mass/PDG from mcparticles, 0 (unidentified) in case the matched particle is charged.
-            const auto mc = (*mcparticles)[mcID];
-            const double mass = (!mc.getCharge()) ? mc.getMass() : 0;
-            const int32_t pdg = (!mc.getCharge()) ? mc.getPDG() : 0;
-            if (m_log->level() <= spdlog::level::debug) {
-                if (mc.getCharge()) {
-                    m_log->debug("   --> associated mcparticle is not a neutral (PDG: {}), "
-                                 "setting the reconstructed particle ID to 0 (unidentified)", mc.getPDG());
-                }
-                m_log->debug("   --> found matching associated mcparticle with PDG: {}, energy: {}", pdg,
-                             mc.getEnergy());
-            }
-
-            // Reconstruct our neutrals and add them to the list
-            const auto outpart = reconstruct_neutral(&clus, mass, pdg);
-            m_log->debug(" --> Reconstructed neutral particle with PDG: {}, energy: {}", outpart.getPDG(),
-                         outpart.getEnergy());
+        trace("    --> Found particle with mcID {}", mcID);
 
-            outparts->push_back(outpart);
+        if (mcID < 0) {
+            debug("    --> cannot match track without associated mcID");
+            continue;
+        }
 
-            // Create truth associations
-            auto assoc = outpartsassoc->create();
-            assoc.setRecID(outpart.getObjectID().index);
-            assoc.setSimID(mcID);
-            assoc.setWeight(1.0);
-            assoc.setRec(outpart);
-            assoc.setSim((*mcparticles)[mcID]);
+        if (clusterMap.count(mcID)) {
+            const auto &clus = clusterMap[mcID];
+            debug("    --> found matching cluster with energy: {}", clus.getEnergy());
+            debug("    --> adding cluster to reconstructed particle");
+            outpart.addToClusters(clus);
+            clusterMap.erase(mcID);
         }
```

Comment:
Might there not be more than one inpart with the same mcID? Similar to how you have delt with the largest cluster energy, is there anything which might be more appropriate than coming from the order in the collection for the tracks.

Reply:
The algorithm can certainly be improved, and this may indeed be one area for improvement :-)

This PR is just an interface change, and assessing the correctness of the interface change by verifying that we get identical results relies on not changing how the algorithm works. So, for changes to the algorithm itself I'd prefer to postpone that to a different PR.

Reply:
Capybara looked fine then to best of my knowledge and the interface changes looked sensible.